File: nodejs/examples/products/v1/insert_product_input_async_sample.js
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// [START merchantapi_insert_product_input_async]
'use strict';
const fs = require('fs');
const authUtils = require('../../authentication/authenticate.js');
const {
  ProductInputsServiceClient,
} = require('@google-shopping/products').v1beta;

/**
 * This class demonstrates how to insert a product input asynchronously.
 */

/**
 * Helper function to generate a random string for offerId
 * @returns {string} A sample offerId.
 */
function generateRandomString() {
  const characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let result = '';
  const length = 8;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * characters.length));
  }
  return result;
}

/**
 * Helper function to create a sample ProductInput object
 * @returns {!object} A sample ProductInput object.
 */
function createRandomProduct() {
  const price = {
    amount_micros: 33450000, // 33.45 USD
    currency_code: 'USD',
  };

  const shipping = {
    price: price,
    country: 'GB',
    service: '1st class post',
  };

  const shipping2 = {
    price: price,
    country: 'FR',
    service: '1st class post',
  };

  const attributes = {
    title: 'A Tale of Two Cities',
    description: 'A classic novel about the French Revolution',
    link: 'https://exampleWebsite.com/tale-of-two-cities.html',
    image_link: 'https://exampleWebsite.com/tale-of-two-cities.jpg',
    availability: 'in stock',
    condition: 'new',
    google_product_category: 'Media > Books',
    gtins: ['9780007350896'],
    shipping: [shipping, shipping2],
    // Price is nested within attributes in the ProductInput message
    price: price,
  };

  // Construct the ProductInput object
  const productInput = {
    contentLanguage: 'en',
    feedLabel: 'CH',
    offerId: generateRandomString(),
    productAttributes: attributes,
  };

  return productInput;
}

/**
 * Inserts multiple product inputs asynchronously.
 * @param {!object} config - Configuration object.
 * @param {string} dataSource - The data source name.
 */
async function asyncInsertProductInput(config, dataSource) {
  // Read merchant_id from the configuration file.
  const merchantInfo = JSON.parse(
    fs.readFileSync(config.merchantInfoFile, 'utf8')
  );
  const merchantId = merchantInfo.merchantId;

  // Construct the parent resource name string.
  const parent = `accounts/${merchantId}`;

  // Get OAuth2 credentials.
  const authClient = await authUtils.getOrGenerateUserCredentials();

  // Create client options with authentication.
  const options = {authClient: authClient};

  // Create the ProductInputsServiceClient.
  const productInputsServiceClient = new ProductInputsServiceClient(options);

  // Create five insert product input requests with random product details.
  const requests = [];
  for (let i = 0; i < 5; i++) {
    const request = {
      parent: parent,
      // You can only insert products into datasource types of Input "API" and "FILE", and
      // of Type "Primary" or "Supplemental."
      // This field takes the `name` field of the datasource, e.g.,
      // accounts/123/dataSources/456
      dataSource: dataSource,
      // If this product is already owned by another datasource, when re-inserting, the
      // new datasource will take ownership of the product.
      productInput: createRandomProduct(),
    };
    requests.push(request);
  }

  console.log('Sending insert product input requests...');

  // Create an array of promises by calling the insertProductInput method for each request.
  const insertPromises = requests.map(request =>
    productInputsServiceClient.insertProductInput(request)
  );

  // Wait for all insert operations to complete.
  // Promise.all returns an array of results, where each result is the response
  // from the corresponding insertProductInput call (which is the inserted ProductInput).
  // The response from insertProductInput is an array where the first element is the ProductInput.
  const results = await Promise.all(insertPromises);
  const insertedProducts = results.map(result => result[0]); // Extract ProductInput from each response array

  console.log('Inserted products below:');
  console.log(JSON.stringify(insertedProducts, null, 2));
}

/**
 * Main function to call the async insert product input method.
 */
async function main() {
  // Get configuration settings.
  const config = authUtils.getConfig();
  // Define the data source ID. Replace {datasourceId} with your actual data source ID.
  // The format is accounts/{account_id}/dataSources/{datasource_id}.
  const merchantInfo = JSON.parse(
    fs.readFileSync(config.merchantInfoFile, 'utf8')
  );
  const merchantId = merchantInfo.merchantId;
  const dataSource = `accounts/${merchantId}/dataSources/{datasourceId}`; // Replace {datasourceId}

  try {
    await asyncInsertProductInput(config, dataSource);
  } catch (error) {
    console.error(`An error occurred: ${error.message || error}`);
    // Log details if available (e.g., for gRPC errors)
    if (error.details) {
      console.error(`Details: ${error.details}`);
    }
  }
}

main();
// [END merchantapi_insert_product_input_async]

File: golang/examples/products/v1/insert_product_input_async_sample.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package v1

// [START merchantapi_insert_product_input_async]

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"sync"
	"time"

	"cloud.google.com/go/shopping/merchant/products/apiv1/productspb"
	"cloud.google.com/go/shopping/type/typepb"
	"google.golang.org/api/option"

	products "cloud.google.com/go/shopping/merchant/products/apiv1"

	"github.com/google/merchant-api-samples/go/collection"
	"github.com/google/merchant-api-samples/go/examples/googleauth"
)

const (
	// The ID of the account that owns the data source.
	accountForInsertAsync = "accounts/1234567890"
	// The ID of the data source that will own the product.
	// You can only insert products into datasource types of Input "API", and of Type
	// "Primary" or "Supplemental."
	// This field takes the `name` field of the datasource.
	dataSourceForInsertAsync = "accounts/1234567890/dataSources/1234567890"
)

var seededRand *rand.Rand = rand.New(
	rand.NewSource(time.Now().UnixNano()))

const charset = "abcdefghijklmnopqrstuvwxyz" +
	"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

// generateRandomString creates a random string of a given length.
func generateRandomString(length int) string {
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[seededRand.Intn(len(charset))]
	}
	return string(b)
}

// createRandomProductInput creates a new ProductInput with a random offer ID.
func createRandomProductInput() *productspb.ProductInput {
	price := &typepb.Price{
		AmountMicros: collection.Ptr(int64(33450000)),
		CurrencyCode: collection.Ptr("USD"),
	}

	shipping1 := &productspb.Shipping{
		Price:   price,
		Country: "GB",
		Service: "1st class post",
	}

	shipping2 := &productspb.Shipping{
		Price:   price,
		Country: "FR",
		Service: "1st class post",
	}

	attributes := &productspb.ProductAttributes{
		Title:                 collection.Ptr("A Tale of Two Cities"),
		Description:           collection.Ptr("A classic novel about the French Revolution"),
		Link:                  collection.Ptr("https://exampleWebsite.com/tale-of-two-cities.html"),
		ImageLink:             collection.Ptr("https://exampleWebsite.com/tale-of-two-cities.jpg"),
		Availability:          collection.Ptr(productspb.Availability_IN_STOCK),
		Condition:             collection.Ptr(productspb.Condition_NEW),
		GoogleProductCategory: collection.Ptr("Media > Books"),
		Gtins:                 []string{"9780007350896"},
		Shipping:              []*productspb.Shipping{shipping1, shipping2},
	}

	return &productspb.ProductInput{
		ContentLanguage:   "en",
		FeedLabel:         "CH",
		OfferId:           generateRandomString(8),
		ProductAttributes: attributes,
	}
}

type insertProductInputAsyncSample struct{}

func init() {
	if err := collection.Add("products.productinputs.v1.insert_product_input_async", &insertProductInputAsyncSample{}); err != nil {
		log.Fatalf("could not add example: %v", err)
	}
}

func (s *insertProductInputAsyncSample) Description() string {
	return "This sample demonstrates how to insert a product input asynchronously"
}

func (s *insertProductInputAsyncSample) Execute() error {
	ctx := context.Background()

	// Authenticates with Google.
	tokenSource, err := googleauth.AuthWithGoogle(ctx)
	if err != nil {
		return fmt.Errorf("failed to authenticate: %w", err)
	}

	// Creates a new product inputs client.
	productInputsClient, err := products.NewProductInputsClient(ctx, option.WithTokenSource(tokenSource))
	if err != nil {
		return fmt.Errorf("could not create product inputs client: %w", err)
	}
	defer productInputsClient.Close()

	parent := accountForInsertAsync

	// Number of concurrent requests to send.
	const numRequests = 5

	var wg sync.WaitGroup
	resultsChan := make(chan *productspb.ProductInput, numRequests)
	errChan := make(chan error, numRequests)

	fmt.Println("Sending insert product input requests")

	// Launch goroutines to send requests concurrently.
	for i := 0; i < numRequests; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()

			req := &productspb.InsertProductInputRequest{
				Parent:       parent,
				DataSource:   dataSourceForInsertAsync,
				ProductInput: createRandomProductInput(),
			}

			// Calls the API to insert the product input.
			resp, err := productInputsClient.InsertProductInput(ctx, req)
			if err != nil {
				errChan <- fmt.Errorf("could not insert product input: %w", err)
				return
			}
			resultsChan <- resp
		}()
	}

	// Wait for all goroutines to finish and then close the channels.
	go func() {
		wg.Wait()
		close(resultsChan)
		close(errChan)
	}()

	// Collect all errors.
	var errs []error
	for err := range errChan {
		errs = append(errs, err)
	}

	// Collect all successful results.
	var results []*productspb.ProductInput
	for result := range resultsChan {
		results = append(results, result)
	}

	// Handle results and errors after all requests are complete.
	if len(errs) > 0 {
		for _, err := range errs {
			fmt.Println(err)
		}
		return fmt.Errorf("encountered %d errors during async insert", len(errs))
	}

	fmt.Println("Inserted products below")
	fmt.Println(results)

	return nil
}

// [END merchantapi_insert_product_input_async]
